### DISCUSSION TITLE
Colormap issue

### DISCUSSION THREAD

**Discussion by abillits** [developmentseed/titiler#449] [Q&A] [answered]
Using the /cog/tiles/ and /cog/preview/ endpoints, whenever I try adding a color map, I get the following:

{"detail":"arrays used as indices must be of integer (or boolean) type"}

I've tried an existing color map such as:
colormap_name=cfastie

I've also tried a very basic colormap such as (but urlencoded):
colormap={"0": "#000000", "255": "#ffffff"}

Can anyone point out what I'm doing wrong?

**Comment by abillits** [comment]
Btw, if I view the file via the /cog/viewer/ endpoint, the selectable colormaps work just fine.

**Reply by vincentsarago** [reply]
Can you make sure the colormap is JSON encoded 
See https://github.com/developmentseed/titiler/blob/master/src/titiler/core/tests/test_dependencies.py#L63

**Reply by vincentsarago** [reply]
confirmed locally, you need to pass colormap encoded in JSON and then urlencoded

```python
import json
from urllib.parse import urlencode

from titiler.core.dependencies import ColorMapParams
from fastapi import FastAPI, Depends, Query

app = FastAPI()

@app.get("/")
def main(cm=Depends(ColorMapParams)):
    """return cmap."""
    return cm


# OK
client = TestClient(app)
cmap = json.dumps({0: "#000000", 255: "#ffffff"})
response = client.get("/", params={"colormap": cmap})
print(response.json())
>> {'0': [0, 0, 0, 255], '255': [255, 255, 255, 255]}

# NOK
client = TestClient(app)
cmap = json.dumps({0: "#000000", 255: "#ffffff"})
response = client.get(f"/?colormap={cmap}")
print(response.json())
>> {'detail': 'Could not parse the colormap value.'}


# OK
client = TestClient(app)
cmap = json.dumps({0: "#000000", 255: "#ffffff"})
response = client.get(f"/?{urlencode({'colormap': cmap})}")
print(response.json())
>> {'0': [0, 0, 0, 255], '255': [255, 255, 255, 255]}
```

**Comment by vincentsarago** [comment, marked as answer]
🤦 I think I miss understood the initial question @idtstudios, 

is it possible that your input data is not of type `Byte` (meaning with value from 0 to 255). If that's the case, colormap won't work because you first need to rescale your data to `byte` (using the `rescale={min},{max}` parameter), which will create value from 0 to 255. Then you can apply colormap

**Reply by iferencik** [reply]
@vincentsarago , i'll use this discussion to shed some light (for me at least) on the colormap topic.
titleir uses [rio-tiler](https://github.com/cogeotiff/rio-tiler), and  [rio-tiler-docs](https://cogeotiff.github.io/rio-tiler/colormap/)  docs indicate there are two types of colormaps

1. **continuous** or **linear** that require the input data to be rescaled to 0-255. I believe titiler uses the colormap_name parameter to
specify a predefined continuous/linear colormaps from [here](https://github.com/cogeotiff/rio-tiler/tree/master/rio_tiler/cmap_data).

This colormaps can be passed to titiler **tiles** endpoint and if the data is  in 0-255 range or the data is  rescaled to this range
the pixels are coloured according to the corresponding color value extracted from interpolating the supplied colormap
in the range (0-255). I have tested these two forms of colormaps without errors

```
{
  value1: (r, g, b, a),
  value2: (r, g, b, a),
}

{ 
                "0": "#7b3294",
                 "73": "#c2a5cf",
                "125": "#f7f7f7",
                "212": "#a6dba0",
                "255": "#008837",
}

{
                 0: [123,50,148],
                  73: [194,165,207],
                 125: [247,247,247],
                 212: [166,219,160],
                 255: [0,136,55],
}
```

2. **discrete** or **interval** based colormaps that can be used for float data as well and where the pixel values are binned according to the min,max values specified  for each class and assigned the corresponding color associated with that class

```
{
  (min, max),(r, g, b, a)),
   (min, max),(r, g, b, a))
}

```
However, I am not able to use successfully the binned/discrete form of the colormaps with titiler.
 @vincentsarago do you mind providing us with a very simple example here? (thank you in advance)




There are two aspects related to handling raster colormaps in a web mapping app that I am interested in. First is of course being able to apply/change  the graphical appearance of a raster tile layer by manipulating the colormaps in conjunction with rescaling and min/max values of the data and second is to build legends using SVG  based legends (linear gradient interpolation for continuous colormaps and simple HTML|SVG for the interval/discrete color maps.

**Reply by vincentsarago** [reply]
`interval` colormap is tested here https://github.com/developmentseed/titiler/blob/master/src/titiler/core/tests/test_factories.py#L134-L149 

I've also updated the titiler notebook: https://github.com/developmentseed/titiler/blob/33d3492551aa6198446dc503842ee6c4e6a31332/docs/examples/notebooks/Working_with_CloudOptimizedGeoTIFF_simple.ipynb to show how to work with rescaling/colormap

**Comment by abillits** [comment]
Ah! That was what I was doing wrong. I'm so used to using palettes with gdal that I didn't even notice that. Thank you!

I can now easy apply one of the premade colormaps (viridis, etc). However, I am now having a bit of trouble with applying my own.

If I provide the following:
{"0":"#ff0000","127":"#00ff00","255":"#0000ff"}

value 0, 127, and 255 have colors applied but none of the in between values do. I have tried adding the following but they have effect:
&colormap_type=linear
&colormap_type=discrete

**Reply by vincentsarago** [reply]
@idtstudios we don't support non-full colormap by default but you can do as explained in https://developmentseed.org/titiler/examples/code/tiler_with_custom_colormap/ 

Basically supporting `colormap_type=linear` needs matplotlib dependency which we didn't want as default dependency.

**Comment by abillits** [comment]
Ah, gotcha. Thank you for taking the time to explain all of this. Much appreciated!
